openassessment/workflow/team_api.py

```diff
@@ -0,0 +1,139 @@
+    try:
+        team_workflow.update_from_assessments()
+        logger.info((
+            u"Updated workflow for team submission UUID {uuid} "
+        ).format(uuid=team_submission_uuid))
+    except Exception as exc:
+        err_msg = (
+            u"Could not update team assessment workflow: {exc}"
+        ).format(exc=exc)
+        logger.exception(err_msg)
+        raise AssessmentWorkflowInternalError(err_msg)
+
+    # Return serialized workflow object
+    return _serialized_with_details(team_workflow)
+
+
+def _serialized_with_details(team_workflow):
+    data_dict = TeamAssessmentWorkflowSerializer(team_workflow).data
+    data_dict['status_details'] = team_workflow.status_details()
+    return data_dict
+
+
+def _get_workflow_model(team_submission_uuid):
+    """
+    Returns the `TeamAssessmentWorkflow` model associated with the
+    given `team_submission_uuid`.
+    """
+    if not isinstance(team_submission_uuid, str):
+        raise AssessmentWorkflowRequestError("team_submission_uuid must be a string")
+
+    try:
+        team_workflow = TeamAssessmentWorkflow.get_by_team_submission_uuid(team_submission_uuid)
+    except Exception as exc:
+        err_msg = (
+            u"Could not get team assessment workflow with team_submission_uuid {uuid} due to error: {exc}"
+        ).format(uuid=team_submission_uuid, exc=exc)
+        logger.exception(err_msg)
+        raise AssessmentWorkflowInternalError(err_msg)
+
+    if team_workflow is None:
+        err_msg = (
+            u"No team assessment workflow matching team_submission_uuid {uuid}"
+        ).format(uuid=team_submission_uuid)
+        raise AssessmentWorkflowNotFoundError(err_msg)
+
+    return team_workflow
+
+
+def get_status_counts(course_id, item_id):
+    """
+    Count how many team workflows have each status, for a given item in a course.
+    "staff" is the only allowed step for team submissions, so we don't
+    need a `steps` parameter here.
+
+    Keyword Arguments:
+        course_id (unicode): The ID of the course.
+        item_id (unicode): The ID of the item in the course.
+
+    Returns:
+        list of dictionaries with keys "status" (str) and "count" (int)
+
+    Example usage:
+        >>> get_status_counts("course-v1:edX+DemoX+Demo_Course", "peer-assessment-problem")
+        [
+            {"status": "staff", "count": 5},
+            {"status": "waiting", "count": 43},
+            {"status": "done", "count": 12},
+        ]
+    """
+    # The AI status exists for workflow logic, but no student will ever be in
+    # the AI status, so we should never return it.
+    statuses = TeamAssessmentWorkflow.STEPS + TeamAssessmentWorkflow.STATUSES
+    if 'ai' in statuses:
+        statuses.remove('ai')
+
+    return [
+        {
+            "status": status,
+            "count": TeamAssessmentWorkflow.objects.filter(
+                status=status,
+                course_id=course_id,
+                item_id=item_id,
+            ).count()
```

Comment:
You should be able to do this with one query with a Django [annotation](https://docs.djangoproject.com/en/2.0/topics/db/aggregation/#topics-db-aggregation):
```python
from django.db.models import Count
queryset = TeamAssessmentWorkflow.objects.filter(
    status__in=statuses,
    course_id=course_id, 
    item_id=item_id,
).values('status').annotate(count=Count('status'))
return [row for row in queryset]
```

Reply:
This produces slightly different behavior since it won't have counts of zero. Suggestions?

Reply:
In that case, using your original query would make sense. Is your issue analogous to [this](https://stackoverflow.com/questions/51858201/django-annotate-count-include-0-results)?

Reply:
@Andytr1, yes, exactly! We could modify this but, following the non-team example: we want to know the steps/statuses even if they have a count of 0
```
[
  {"staus": "staff", "count": 0}
  {"status": "waiting", "count": 1}
  ...
]
```

Reply:
If I was to implement @iloveagent57's suggestion, one approach is to create a dictionary with all statuses and the count initialized to 0. Then get the results of the query and set the values in the existing dictionary to the actual count. Any statuses not found in db, would remain at 0. It would be a little more code, but only a single query against the db would be executed.

Reply:
Yeah, what Andrew suggested is what I would do, too, something like:
```python
queryset = TeamAssessmentWorkflow.objects.filter(
    status__in=statuses,
    course_id=course_id, 
    item_id=item_id,
).values('status').annotate(count=Count('status'))
counts_by_status = {status: 0 for status in statuses}
for row in queryset:
    counts_by_status[row['status']] = row['count']
return [
    {'status': status, 'count': count}
    for status, count
    in counts_by_status.items()
]
```